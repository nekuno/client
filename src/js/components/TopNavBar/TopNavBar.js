import PropTypes from 'prop-types';
import React, {Component} from 'react';
import styles from './TopNavBar.scss';
import TopBar from '../ui/TopBar/TopBar.js';
import IconNotification from '../ui/IconNotification/IconNotification.js';
import RoundedIcon from '../ui/RoundedIcon/RoundedIcon.js';
import RoundedImage from '../ui/RoundedImage/RoundedImage.js';
import Input from '../ui/Input/Input.js';
import RouterActionCreators from '../../actions/RouterActionCreators';

export default class TopNavBar extends Component {

    static propTypes = {
        position                     : PropTypes.oneOf(['relative', 'absolute']),
        background                   : PropTypes.string,
        color                        : PropTypes.string,
        boxShadow                    : PropTypes.bool,
        textAlign                    : PropTypes.oneOf(['center', 'left']),
        textSize                     : PropTypes.oneOf(['regular', 'small']),
        iconLeft                     : PropTypes.string,
        textLeft                     : PropTypes.string,
        imageLeft                    : PropTypes.string,
        textLeftColored              : PropTypes.bool,
        textCenter                   : PropTypes.string,
        firstIconRight               : PropTypes.string,
        secondIconRight              : PropTypes.string,
        iconsRightBackground         : PropTypes.string,
        iconsRightColor              : PropTypes.string,
        imageRight                   : PropTypes.string,
        proposalsIcon                : PropTypes.bool,
        proposalsCount               : PropTypes.number,
        messagesIcon                 : PropTypes.bool,
        messagesCount                : PropTypes.number,
        textRight                    : PropTypes.string,
        textRightColored             : PropTypes.bool,
        searchInput                  : PropTypes.bool,
        onLeftLinkClickHandler       : PropTypes.func,
        onSearchChange               : PropTypes.func,
        onRightLinkClickHandler      : PropTypes.func,
        onSecondRightLinkClickHandler: PropTypes.func,
        online                       : PropTypes.bool,
        children                     : PropTypes.object,
    };

    static contextTypes = {
        router: PropTypes.object.isRequired
    };

    constructor(props) {
        super(props);

        this.handleSearchClick = this.handleSearchClick.bind(this);
        this.handleChangeInput = this.handleChangeInput.bind(this);

        this.state = {
            searching: false
        };
    }

    handleLeftClick() {
        if (this.props.onLeftLinkClickHandler) {
            this.props.onLeftLinkClickHandler();
        } else {
            this.goBack.bind(this)();
        }
    }

    handleSearchClick() {
        this.setState({searching: true});
    }

    handleChangeInput(value) {
        if (!value) {
            this.setState({searching: false});
        } else {
            this.props.onSearchChange(value);
        }
    }

    goBack() {
        RouterActionCreators.previousRoute(this.context.router.getCurrentLocation().pathname || '');
    }

    goToProposals() {

    }

    goToMessages() {

    }

    render() {
        const {
            position,
            background,
            color,
            boxShadow,
            textAlign,
            textSize,
            iconLeft,
            textLeft,
            imageLeft,
            textLeftColored,
            textCenter,
            firstIconRight,
            secondIconRight,
            iconsRightBackground,
            iconsRightColor,
            imageRight,
            proposalsIcon,
            proposalsCount,
            messagesIcon,
            messagesCount,
            textRight,
            textRightColored,
            searchInput,
            online,
            children
        } = this.props;
        const {searching} = this.state;
        const leftClassName = textSize === 'small' ? styles.left + ' ' + styles.small : styles.left;
        const centerClassName = textSize === 'small' ? styles.center + ' ' + styles.small : styles.center;
        const rightClassName = textSize === 'small' ? styles.right + ' ' + styles.small : styles.right;
        const textLeftClassName = textLeftColored ? styles.linkColored + ' ' + styles.link : styles.link;
        const textRightClassName = textRightColored ? styles.linkColored + ' ' + styles.link : styles.link;

        return (
            <TopBar position={position} background={background} textAlign={textAlign} color={color} boxShadow={boxShadow}>
                {searching ?
                    <div className={styles.searchInput}>
                        <Input placeholder={''} searchIcon={true} size={'small'} onChange={this.handleChangeInput}/>
                    </div>
                    :
                    <div>
                        <div className={leftClassName} onClick={this.handleLeftClick.bind(this)}>
                            {iconLeft ?
                                <div className={styles.icon + ' icon icon-' + iconLeft}/>
                                : imageLeft ?
                                    <RoundedImage url={imageLeft}
                                                  size="x-small"
                                    />
                                    : <div className={textLeftClassName}>{textLeft}</div>
                            }
                        </div>

                        <div className={centerClassName}>
                            {textCenter || ' '}
                        </div>

                        <div className={rightClassName}>
                            {searchInput ?
                                <div className={styles.icon + ' ' + styles.iconSearch + ' icon icon-search'} onClick={this.handleSearchClick}/>
                                : null
                            }
                            {firstIconRight === 'icon-proyecto' || firstIconRight === 'icon-hobbie' || firstIconRight === 'icon-experiencia' ?
                                <span className={styles.icon + " " + styles.iconProposals + " " + firstIconRight}>
                                    <span className="path1"></span>
                                    <span className="path2"></span>
                                    <span className="path3"></span>
                                    <span className="path4"></span>
                                    <span className="path5"></span>
                                    <span className="path6"></span>
                                    <span className="path7"></span>
                                </span>
                                : null
                            }
                            {iconsRightBackground ?
                                <div className={styles.roundedIcon}>
                                    <RoundedIcon icon={firstIconRight}
                                                 size="small"
                                                 background={iconsRightBackground}
                                                 color={iconsRightColor}
                                                 fontSize={'15px'}
                                                 onClickHandler={this.props.onRightLinkClickHandler}
                                    />
                                </div>
                                :
                                    <div className={styles.icon + ' icon icon-' + firstIconRight} style={{color: iconsRightColor}} onClick={this.props.onRightLinkClickHandler}/>
                            }
                            {secondIconRight ?
                                iconsRightBackground ?
                                    <div className={styles.roundedIcon + ' ' + styles.second}>
                                        <RoundedIcon icon={secondIconRight}
                                                     size="small"
                                                     background={iconsRightBackground}
                                                     color={iconsRightColor}
                                                     fontSize={'15px'}
                                                     onClickHandler={this.props.onSecondRightLinkClickHandler}
                                        />
                                    </div>
                                    :
                                    <div className={styles.icon + ' icon icon-' + secondIconRight + ' ' + styles.second} style={{color: iconsRightColor}} onClick={this.props.onSecondRightLinkClickHandler}/>
                                : null
                            }
                            {imageRight ?
                                <div>
                                    <RoundedImage url={imageRight}
                                                  size="small"
                                                  onClickHandler={this.props.onRightLinkClickHandler}
                                    />
                                    {online !== undefined ?
                                        online ?
                                        <div className={styles.online}></div>
                                            : <div className={styles.offline}></div>
                                        : null
                                    }
                                </div>
                                : null
                            }
                            {messagesIcon ?
                                <IconNotification icon={'mail'}
                                                  notifications={messagesCount}
                                                  onClickHandler={this.goToMessages.bind(this)}
                                />
                                : null
                            }
                            {proposalsIcon ?
                                <IconNotification icon={'user'}
                                                  notifications={proposalsCount}
                                                  onClickHandler={this.goToProposals.bind(this)}
                                />
                                : null
                            }
                            {textRight ?
                                <div className={textRightClassName} onClick={this.props.onRightLinkClickHandler}>
                                    {textRight}
                                </div>
                                : null
                            }
                        </div>
                        {children}
                    </div>
                }
            </TopBar>
        );
    }
}