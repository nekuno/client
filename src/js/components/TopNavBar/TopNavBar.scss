@import '../../../scss/_partials/colors';

.search-input {
  margin-top: -20px;
  position: absolute;
  width: 100%;
}

.left {
  cursor: pointer;
  display: inline-block;
  height: 50px;
  padding: 0 2%;
  text-align: left;
  vertical-align: top;
  width: 21%;

  .icon {
    display: inline-block;
    font-size: 25px;
    line-height: 50px;
    margin-top: 0;
    text-align: center;
    height: 50px;
    width: 40px;
  }
  .rounded-icon {
    display: inline-block;
  }

  &.small {
    .icon {
      margin-top: 8px;
    }
  }

}
.center {
  display: inline-block;
  height: 100%;
  width: 50%;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  vertical-align: top;

  &.small {
    font-size: 13px;
    line-height: 40px;
    vertical-align: text-top;
  }
}
.right {
  display: inline-block;
  height: 50px;
  padding: 0 2%;
  text-align: right;
  vertical-align: top;
  width: 21%;

  .icon {
    cursor: pointer;
    display: inline-block;
    font-size: 25px;
    line-height: 50px;
    margin-top: 0;
    text-align: center;
    height: 50px;
    width: 40px;

    &.second {
      margin-left: 20px;
    }
    &.icon-search {
      color: $purple-main;
    }
  }
  .rounded-icon {
    display: inline-block;
    height: 35px;
    padding: 7.5px 0;

    &.second {
      margin-left: 20px;
    }
  }

  &.small {

    .icon {
      margin-top: 8px;
    }
  }
}