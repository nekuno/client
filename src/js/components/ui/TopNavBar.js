import PropTypes from "prop-types";
import React, { Component } from "react";
import shouldPureComponentUpdate from "react-pure-render/function";
import Icon from './Icon';
import AuthenticatedComponent from '../AuthenticatedComponent';
import RouterActionCreators from '../../actions/RouterActionCreators';

function parsePicture(user) {
    return user && user.photo ? user.photo.thumbnail.small : 'img/no-img/small.jpg';
}

@AuthenticatedComponent
export default class TopNavBar extends Component {
	static contextTypes = {
		router: PropTypes.object.isRequired
	};
	static propTypes = {
		leftMenuIcon: PropTypes.bool,
		leftIcon: PropTypes.string,
		centerText: PropTypes.string,
		rightIcon: PropTypes.string,
		onLeftLinkClickHandler: PropTypes.func,
		onRightLinkClickHandler: PropTypes.func,
		wrapIcons: PropTypes.bool,
		background: PropTypes.string,
		color: PropTypes.string,
		// Injected by @AuthenticatedComponent
        user: PropTypes.object.isRequired,
	};

	goBack() {
		RouterActionCreators.previousRoute(this.context.router.getCurrentLocation().pathname);
	}

	shouldComponentUpdate = shouldPureComponentUpdate;

	render() {
		const {
			user,
			leftMenuIcon,
			centerText,
			onLeftLinkClickHandler,
			onRightLinkClickHandler,
			background,
			color,
			wrapIcons,
		} = this.props;
		let navBarClass = "navbar-no-f7";
		if (wrapIcons) navBarClass += ' wrap-icons';
		
		const style = {};
		if (background) {
			style.background = background;
			style.boxShadow = 'none';
		}
		if (color) style.color = color;
		
		// FIXME
		let { leftIcon, rightIcon } = this.props;
		if (leftIcon === 'left-arrow') leftIcon = 'mdi-arrow-left';
		if (leftIcon === 'arrow-left') leftIcon = 'mdi-arrow-left';
		if (leftIcon === 'x') leftIcon = 'mdi-close';
		if (rightIcon === 'x') rightIcon = 'mdi-close';
		if (leftIcon === 'check') leftIcon = 'mdi-check';
		if (rightIcon === 'check') rightIcon = 'mdi-check';

		return (
			<div className={navBarClass} style={style}>
				{leftMenuIcon ? (
					<a className="left open-panel" id="joyride-3-menu">
						<img src={parsePicture(user)} className="profile-image" />
					</a>
				) : leftIcon ? (
					<a className="left" onClick={onLeftLinkClickHandler || this.goBack.bind(this)}>
						<span className="navbar-icon">
							<Icon icon={leftIcon} />
						</span>
					</a>
				) : <span className="left-gap"></span> }
				
				<span className="center">
					{ centerText ? <span className="text">{centerText}</span> : '' }
				</span>
				
				{ rightIcon ? (
					<a className="right" onClick={onRightLinkClickHandler || this.goBack.bind(this)}>
						<span className="navbar-icon">
							<Icon icon={rightIcon} />
						</span>
					</a>
				) : <span className="right"></span> }
			</div>
		);
	}
}
