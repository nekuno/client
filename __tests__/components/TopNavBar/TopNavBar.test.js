import React from 'react';
import TopNavBar from '../../../src/js/components/TopNavBar/TopNavBar.js';
import styles from '../../../src/js/components/TopNavBar/TopNavBar.scss';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

describe('Test TopNavBar component', () => {
    const leftClickMockFn = jest.fn();
    const rightClickMockFn = jest.fn();
    const context = { router: { getCurrentLocation: () => ({pathname: 'foo'}) } };
    const topNavBar = shallow(
        <TopNavBar leftIcon={'left-arrow'}
                   onLeftLinkClickHandler={leftClickMockFn}
                   textCenter={'Foo'}
                   firstIconRight={'delete'}
                   onRightLinkClickHandler={rightClickMockFn}
        />, {context}
    );

    it('should be defined', () => {
        expect(TopNavBar).toBeDefined();
    });
    it('should render correctly', () => {
        expect(topNavBar).toMatchSnapshot();
    });
    it('TopNavBar left click event', () => {
        topNavBar.find(`.${styles.left}`).simulate('click');
        expect(leftClickMockFn).toHaveBeenCalledTimes(1);
    });
    it('TopNavBar right click event', () => {
        topNavBar.find(`.${styles.right}` + ' ' + `.${styles.icon}`).simulate('click');
        expect(rightClickMockFn).toHaveBeenCalledTimes(1);
    });
});
