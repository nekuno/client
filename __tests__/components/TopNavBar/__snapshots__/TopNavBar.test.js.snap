// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test TopNavBar component should render correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TopNavBar
    firstIconRight="delete"
    leftIcon="left-arrow"
    onLeftLinkClickHandler={[MockFunction]}
    onRightLinkClickHandler={[MockFunction]}
    textCenter="Foo"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <div
          className="left"
          onClick={[Function]}
        >
          <div
            className="link"
          />
        </div>,
        <div
          className="center"
        >
          Foo
        </div>,
        <div
          className="right"
        >
          <div
            className="icon icon icon-delete"
            onClick={[MockFunction]}
          />
        </div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div
            className="link"
          />,
          "className": "left",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": undefined,
            "className": "link",
          },
          "ref": null,
          "rendered": null,
          "type": "div",
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Foo",
          "className": "center",
        },
        "ref": null,
        "rendered": "Foo",
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="icon icon icon-delete"
              onClick={[MockFunction]}
            />,
            null,
            null,
            null,
            null,
            null,
          ],
          "className": "right",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "icon icon icon-delete",
              "onClick": [MockFunction],
            },
            "ref": null,
            "rendered": null,
            "type": "div",
          },
          null,
          null,
          null,
          null,
          null,
        ],
        "type": "div",
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <div
            className="left"
            onClick={[Function]}
          >
            <div
              className="link"
            />
          </div>,
          <div
            className="center"
          >
            Foo
          </div>,
          <div
            className="right"
          >
            <div
              className="icon icon icon-delete"
              onClick={[MockFunction]}
            />
          </div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div
              className="link"
            />,
            "className": "left",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": undefined,
              "className": "link",
            },
            "ref": null,
            "rendered": null,
            "type": "div",
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Foo",
            "className": "center",
          },
          "ref": null,
          "rendered": "Foo",
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="icon icon icon-delete"
                onClick={[MockFunction]}
              />,
              null,
              null,
              null,
              null,
              null,
            ],
            "className": "right",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "icon icon icon-delete",
                "onClick": [MockFunction],
              },
              "ref": null,
              "rendered": null,
              "type": "div",
            },
            null,
            null,
            null,
            null,
            null,
          ],
          "type": "div",
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "context": Object {
      "router": Object {
        "getCurrentLocation": [Function],
      },
    },
  },
}
`;
